sql: Fix virtual index scan for crdb_internal.create_type_statements

spilchen opened the pull request:
When querying the crdb_internal.create_type_statements virtual table, two access paths are possible: a full table scan or a virtual index scan. We observed inconsistent behavior between these paths. The full table scan returned correct data, but the virtual index scan could return types for the wrong database (if one was specified) or panic with a null pointer dereference when querying across all databases. This change corrects the virtual index access pattern to ensure consistency with the full table scan.

Epic: None
Closes #137579
Release note (bug fix): Fixed issues with the virtual index scan on crdb_internal.create_type_statements, ensuring consistent results when querying user-defined types (UDTs) across databases.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/138295)
<!-- Reviewable:end -->

rafiss reviewed (approved):
nice fix!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138295#-:-OFwIbgR1z5X7jFvWBaM:bhpddey)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/logictest/testdata/logic_test/crdb_internal` line 1700 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/138295#-OFwHyvY9jVWVrpfptkB:-OFwHyvZ3goGuKxYTKvn:b-i1o3ll) ([raw file](https://github.com/cockroachdb/cockroach/blob/513301ddccdd96b69064f790652bd08ba26f8018/pkg/sql/logictest/testdata/logic_test/crdb_internal#L1700)):*
> ```
> # This one uses the full table scan.
> query ITTITTT
> SELECT * FROM crdb_internal.create_type_statements WHERE descriptor_id = (('other_db.public.enum1'::regtype::int) - 100000)::oid
> ```

super nit: a more reliable way to force a full table scan might be to use the name in the where clause:
```
SELECT * FROM crdb_internal.create_type_statements WHERE descriptor_name = 'enum1'
```

I think it might be better since in the future the optimizer might be smart enough to use the index for the int-oid comparison.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138295)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/logictest/testdata/logic_test/crdb_internal` line 1700 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/138295#-OFwHyvY9jVWVrpfptkB:-OFwZD0TE-OQh20dj3cF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/513301ddccdd96b69064f790652bd08ba26f8018/pkg/sql/logictest/testdata/logic_test/crdb_internal#L1700)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: a more reliable way to force a full table scan might be to use the name in the where clause:
```
SELECT * FROM crdb_internal.create_type_statements WHERE descriptor_name = 'enum1'
```

I think it might be better since in the future the optimizer might be smart enough to use the index for the int-oid comparison.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql: Fix virtual index scan for crdb_internal.create_type_statements

When querying the crdb_internal.create_type_statements virtual table,
two access paths are possible: a full table scan or a virtual index
scan. We observed inconsistent behavior between these paths. The full
table scan returned correct data, but the virtual index scan could
return types for the wrong database (if one was specified) or panic with
a null pointer dereference when querying across all databases. This
change corrects the virtual index access pattern to ensure consistency
with the full table scan.

Epic: None
Closes #137579
Release note (bug fix): Fixed issues with the virtual index scan on
crdb_internal.create_type_statements, ensuring consistent results when
querying user-defined types (UDTs) across databases.

spilchen commented:
TFTR!

bors r+
